File: 0394-decode-string/0394-decode-string.py
class Solution:
    def decodeString(self, s: str) -> str:
        count_stack = []
        result_stack = []
        current_string = []
        k = 0

        for ch in s:
            if ch.isdigit():
                k = k * 10 + int(ch)
            elif ch == '[':
                count_stack.append(k)
                result_stack.append(''.join(current_string))
                k = 0
                current_string = []
            elif ch == ']':
                repeat_times = count_stack.pop()
                prev_string = result_stack.pop()
                current_string = [prev_string + ''.join(current_string) * repeat_times]
            else:
                current_string.append(ch)

        return ''.join(current_string)

File: 0394-decode-string/0394-decode-string.ts
function decodeString(s: string): string {
    const countStack: number[] = [];
    const resultStack: string[] = [];
    let currentString: string[] = [];
    let k = 0;

    for (const ch of s) {
        if (/\d/.test(ch)) {
            k = k * 10 + parseInt(ch, 10);
        } else if (ch === '[') {
            countStack.push(k);
            resultStack.push(currentString.join(''));
            k = 0;
            currentString = [];
        } else if (ch === ']') {
            const repeatTimes = countStack.pop()!;
            const prevString = resultStack.pop()!;
            currentString = [(prevString + currentString.join('').repeat(repeatTimes))];
        } else {
            currentString.push(ch);
        }
    }

    return currentString.join('');
}
